File: Data Structure & Algorithm/Template/Binary Tree/Binary Tree Traversal/Pre-order/Recursion/Preorder.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

export function preorderTraversalRecursive(root: TreeNode | null): number[] {
  const res: number[] = [];

  const recursive = (node: TreeNode | null) => {
    if (!node) {
      return;
    } else {
      res.push(node.val);
      recursive(node?.left);
      recursive(node?.right);
    }
  };

  recursive(root);

  return res;
}

// test
const root: TreeNode = {
  val: 1,
  left: {
    val: 2,
    left: {
      val: 4,
      left: null,
      right: null,
    },
    right: {
      val: 5,
      left: null,
      right: null,
    },
  },
  right: {
    val: 3,
    left: {
      val: 6,
      left: {
        val: 7,
        left: null,
        right: null,
      },
      right: null,
    },
    right: null,
  },
};
const res = preorderTraversalRecursive(root);
console.log(res);

File: Data Structure & Algorithm/Template/Binary Tree/Binary Tree Traversal/Pre-order/Recursion/Preorder.go
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func preorderTraversalRecursive(root *TreeNode) []int {
	res := []int{}

	var recursive func(node *TreeNode)

	recursive = func(node *TreeNode) {
		if node == nil {
			return
		}

		res = append(res, node.Val)
		recursive(node.Left)
		recursive(node.Right)
	}

	recursive(root)
	return res
}

// test
func main() {
	root := &TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 4,
			},
			Right: &TreeNode{
				Val: 5,
			},
		},
		Right: &TreeNode{
			Val: 3,
			Left: &TreeNode{
				Val: 6,
				Right: &TreeNode{
					Val: 7,
				},
			},
		},
	}

	res := preorderTraversalRecursive(root)
	fmt.Println(res)
}
